src/lib/dnssd/IncrementalResolve.h

```diff
@@ -0,0 +1,168 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/dnssd/Resolver.h>
+#include <lib/dnssd/minimal_mdns/Parser.h>
+#include <lib/dnssd/minimal_mdns/RecordData.h>
+#include <lib/dnssd/minimal_mdns/core/QName.h>
+#include <lib/support/BitFlags.h>
+#include <lib/support/Variant.h>
+
+namespace chip {
+namespace Dnssd {
+
+/// Allows storing and retrieving of a SerializedQName for later recall.
+///
+/// This is a convenience storage as MDNS generally provides data as QNames
+/// and comparisons between QNames is more convenient than replacing them with
+/// null-terminated character strings.
+class StoredServerName
+{
+public:
+    StoredServerName() {}
+
+    void Clear() { memset(mBuffer, 0, sizeof(mBuffer)); }
+
+    /// Set the underlying value. Will return CHIP_ERROR_NO_MEMORY
+    /// on insufficient storage space.
+    ///
+    /// If insufficient space, buffer will be cleared.
+    CHIP_ERROR Set(mdns::Minimal::SerializedQNameIterator value);
+
+    /// Return the underlying value in this object.
+    ///
+    /// Value valid as long as this object is valid and underlying Set() is
+    /// not called.
+    mdns::Minimal::SerializedQNameIterator Get() const;
+
+private:
+    // Try to have space for at least:
+    //  L1234._sub._matterc._udp.local       => 30 chars
+    //  <fabric>-<node>._mattrer._tcp.local  => 52 chars
+    //  <hostname>.local (where hostname is kHostNameMaxLength == 16)
+    //
+    // This does not try to optimize out ".local" suffix which is always expected
+    // since comparisons are easier when suffix is still present.
+    static constexpr size_t kMaxStoredServerNameLength = 64;
+
+    uint8_t mBuffer[kMaxStoredServerNameLength] = {};
+};
+
+/// Incrementally accumulates data from DNSSD packets. It is generally geared
+/// towards processing minmdns records.
```

Comment:
"towards processing minmdns records" has too much context for someone unaware to the different backends. Suggest rewording as "it is geared towards resource-constrained DNS-SD querier implementations, especially for mDNS"